### PR TITLE
FormCommit: Add soft reset for amending

### LINKED ISSUE #2260: amend should show files of amended commit
Just like git gui, git extensions should not only show the commit message of the amended commit, but also the affected files (patches). Right now I can only add additional changes when amending, but cannot unstage anything from the last commit. I always have to use git gui for that.

### PR DESCRIPTION
**Pull request by mstv**
Fixes #2260

## Proposed changes

- Add `Reset soft` button for amending a commit
- Disable the `Amend commit` checkbox only after Soft Reset, not in dependency of staging
- `ReferenceRepository`: Write git index (LibGitSharp does not do this automatically)
- BugFix: Update button states after commit (`Reset.Enabled` and `CommitAndPush.Text`)

### Before

image

### After

image

ResetSoft_twice

## Test methodology 

- add testcase for workflow
- adapt other testcases to `ReferenceRepository`

## Test environment(s) 

- Git Extensions 33.33.33
- Build a5b0156026f69e11a2876d94a79f012807cf0d39
- Git 2.39.2.windows.1
- Microsoft Windows NT 10.0.19045.0
- .NET 6.0.14
- DPI 96dpi (no scaling)
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by RussKie — COMMENTED** [review]
I'm not fond of adding more controls to this panel - it is already too "bisy". Other than that, it looks ok.

**Comment by mstv** [comment]
> I'm not fond of adding more controls to this panel - it is already too "busy". Other than that, it looks ok.

This (checkbox and) button are invisible until `Amend commit` is ticked.
I'd like to merge it tomorrow.

**Review by gerhardol — COMMENTED** [review]
+1
Agree to this behavior, something "virtual" showing the complete change without an actual reset is too complicated.
Very brief review though. The enabling of Amend seem to be the change to focus on...

**Comment by mstv** [comment]
When amending by means of Soft Reset, the changes need to be committed again before another amend makes sense.

ResetSoft_twice

**Comment by RussKie** [comment]
Let's take it! I will annoy you with a UI layout change later 😆

**Review by RussKie — APPROVED** [review]
Approving but it'd be great to add some tests.

**Comment by mstv** [comment]
>  I will annoy you with a UI layout change later 😆

Why do you expect this? Are you going to break the workflow `Alt+A`, `Alt+F`, `Ctrl+Enter` / `Alt+P`?

> but it'd be great to add some tests

I came to the same conclusion when I was testing it manually.

The `ReferenceRepository` is driving me nuts though. It seems that `repository.Index.Add(_fileName);` does not really stage the file as expected - but just internally.
Then `Commit(repository, commitMessage);` creates a commit with the file. But the staged state of the file becomes "deleted".
This confuses me and my test because the repo is not clean. It is not easily possible to bring the repo into the clean state using the `ReferenceRepository`. Any clues?

image

**Comment by RussKie** [comment]
We do have some existing test for the Commit dialog, where (iirc) we stage
and unstage files. Did you check those?

I can try to help you, if you have a failing test, but I can't promise any
timeline atm - my work keeps me really busy...

**Comment by mstv** [comment]
One needs to call `repository.Index.Write();` in order to let LibGitSharp update the `.git` folder.
Unfortunately, this breaks other tests, which rely on the weird not updated git index.

**Comment by RussKie** [comment]
I'm kind of wondering - do we need the "soft reset" button at all? Shouldn't we just execute it _when_ "amend" is ticked? That is, if a user chose to amend, it's logical to assume he/she'd want to see the amended files? Or am I overthinking it?

The other question I have (I apologise, I haven't had a chance to try this branch) is about the experience when the user closes the dialog after soft resetting and not committing the changes, or simply changes the mind about amending. What are we doing with the soft-reset files?

File: IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormCommitTests.cs [hunk 1]
```diff
@@ -419,6 +428,77 @@ public void ResetAuthor_depends_on_amend()
             });
         }
 
+        [Test]
+        public void ResetSoft()
```
Comment (RussKie) [inline IntegrationTests/UI.IntegrationTests/CommandsDialogs/FormCommitTests.cs:432]: 👍

**Comment by mstv** [comment]
> Shouldn't we just execute it when "amend" is ticked?

You have already answered it yourself: "... changes the mind about amending. What are we doing with the soft-reset files?"
And after soft-reset, one does not see any longer what is being changed, i.e. the diff to the commit being amended.
That's why the removal of the previous commit should be a conscious decision by the user, who must take the responsibility for the resulting files, too.

**Comment by gerhardol** [comment]
I often toggle amend box to get the commit message to edit it for a squash (or amend! now when I know about it) (instead of using ctrl-x and changing the message)

**Comment by RussKie** [comment]
I do check the box a lot, similar to @gerhardol.

I wonder if this functionality can be approached a little differently. That is, when the box is checked we automatically read HEAD~1 content and display it in the "staged" area. This would allow to review the content. These files can be marked differently - e.g., they can't be immediately unstaged or reset.
An attempt to unstage or reset would show a confirmation dialog - "Do you want to perform soft reset?", and if yes, we perform a soft reset and allow interact with all files.
This is not an attempt to derail or block this implementation, and more an exploration of whether we can provide a more seamless experience.

**Comment by gerhardol** [comment]
Showing head~1 in staged would be confusing. A third read only file list could be shown but that will still not show all changes to a file in this list. I often use browse for such manipulations when i do not want to amend just getting an overview for a fixup.

Soft reset is good enough for now.

**Comment by vbjay** [comment]
What if they reset soft and then change their minds?  Uncheck ammend after soft reset?

**Comment by mstv** [comment]
> What if they reset soft and then change their minds? Uncheck amend after soft reset?

They need to commit again (or use the reflog).
The `Amend` checkbox is unticked and disabled on soft reset.

**Comment by vbjay** [comment]
>The Amend checkbox is unticked and disabled on soft reset.

Till they open commit window again?  They could merge commits by:

1. Amend
1. Soft reset
1. Close commit window
1. Open commit window
1. Amend
1. Soft reset
1. Wash..rinse repeat.

**Comment by mstv** [comment]
This soft reset needs three clicks by default, Jay. We cannot take responsibility of all user clicks. The reflog is still there.

One can squash commits using this with reopening FormCommit. :)

**Comment by vbjay** [comment]
Yep. More documenting exactly that in the ticket so when someone comes
around
....but why...how do I...  Already in ticket.

**Comment by vbjay** [comment]
One more. Have you tried with a skip work tree file that is in commit?

**Comment by mstv** [comment]
> One more. Have you tried with a skip work tree file that is in commit?

I have not. But very likely, the file will not be committed as git was told so.
Forgetting about skip-worktree files is a trap.
Though in the case of amending, it should not be long ago.

**Pull request merged**